consumer_test.go

```diff
@@ -56,195 +53,491 @@ func TestConsumerOffsetManual(t *testing.T) {
+	// Given
+	broker0 := newMockBroker(t, 0)
+	defer broker0.Close()
+
+	broker0.SetHandlerByMap(map[string]MockResponse{
+		"MetadataRequest": newMockMetadataResponse(t).
+			SetBroker(broker0.Addr(), broker0.BrokerID()).
+			SetLeader("my_topic", 0, broker0.BrokerID()),
+		"OffsetRequest": newMockOffsetResponse(t).
+			SetOffset("my_topic", 0, OffsetOldest, 0).
+			SetOffset("my_topic", 0, OffsetNewest, 1000),
+		"FetchRequest": newMockFetchResponse(t, 1),
+	})
 
-	// we deliver one message, so it should be one higher than we return in the OffsetResponse
-	if msg.Offset != 10 {
-		t.Error("Latest message offset not fetched correctly:", msg.Offset)
+	config := NewConfig()
+	config.ChannelBufferSize = 0
+	c, err := NewConsumer([]string{broker0.Addr()}, config)
+	if err != nil {
+		t.Fatal(err)
 	}
 
-	if hwmo := consumer.HighWaterMarkOffset(); hwmo != 14 {
-		t.Errorf("Expected high water mark offset 14, found %d", hwmo)
+	pc1, err := c.ConsumePartition("my_topic", 0, 0)
+	if err != nil {
+		t.Fatal(err)
 	}
 
-	leader.Close()
-	safeClose(t, consumer)
-	safeClose(t, master)
+	// When
+	pc2, err := c.ConsumePartition("my_topic", 0, 0)
 
-	// We deliver one message, so it should be one higher than we return in the OffsetResponse.
-	// This way it is set correctly for the next FetchRequest.
-	if consumer.(*partitionConsumer).offset != 11 {
-		t.Error("Latest offset not fetched correctly:", consumer.(*partitionConsumer).offset)
+	// Then
+	if pc2 != nil || err != ConfigurationError("That topic/partition is already being consumed") {
+		t.Fatal("A partition cannot be consumed twice at the same time")
 	}
+
+	safeClose(t, pc1)
+	safeClose(t, c)
 }
 
-func TestConsumerShutsDownOutOfRange(t *testing.T) {
-	seedBroker := newMockBroker(t, 1)
-	leader := newMockBroker(t, 2)
+// If consumer fails to refresh metadata it keeps retrying every with frequency
+// given in `Config.Consumer.Retry.Backoff`.
+func TestConsumerLeaderRefreshError(t *testing.T) {
+	// Given
+	broker0 := newMockBroker(t, 100)
+	defer broker0.Close()
+
+	// Stage 1: my_topic/0 served by broker0
+	Logger.Printf("    STAGE 1")
+
+	broker0.SetHandlerByMap(map[string]MockResponse{
+		"MetadataRequest": newMockMetadataResponse(t).
+			SetBroker(broker0.Addr(), broker0.BrokerID()).
+			SetLeader("my_topic", 0, broker0.BrokerID()),
+		"OffsetRequest": newMockOffsetResponse(t).
+			SetOffset("my_topic", 0, OffsetOldest, 123).
+			SetOffset("my_topic", 0, OffsetNewest, 1000),
+		"FetchRequest": newMockFetchResponse(t, 1).
+			SetMessage("my_topic", 0, 123, testMsg),
+	})
+
+	config := NewConfig()
+	config.Net.ReadTimeout = 100 * time.Millisecond
+	config.Consumer.Retry.Backoff = 500 * time.Millisecond
+	config.Metadata.Retry.Max = 0
+	c, err := NewConsumer([]string{broker0.Addr()}, config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	pc, err := c.ConsumePartition("my_topic", 0, OffsetOldest)
+	if err != nil {
+		t.Errorf("Failed to create a partition consumer, err=%v", err)
```

Comment:
this can be fatal, otherwise the next line is just going to panic with a nil pointer